### ISSUE TITLE
Test Bash Pipelines

### ISSUE THREAD

**Issue by rennergade** [Lind-Project/lind_project#167]
With the recent success with pipes, I wanted to start trying out real bash pipelines.

As an initial test I did `cat 1gbfile | cat > new1gbfile` and found some interesting things. If `new1gbfile` doesn't already exist, Native goes extremely fast, faster than the results I got for even just piping 1GB. I'm assuming there are some FS tricks going on there. If the file does exist, it seems to run at a "normal" pace.

Results:
Native (no file existing): 400ms
Native (file exists prev.): 2400ms
Lind (both): 2100ms

So Lind outperforms native when we don't presumably see some underlying file tricks (I'll need to get down to the bottom of this). I have flamegraphs for all three cases here:
[Native - no prev file](https://github.com/Lind-Project/lind_project/blob/rust-graphs/graphs/catcat-native1.svg)
[Native - prev file exists](https://github.com/Lind-Project/lind_project/blob/rust-graphs/graphs/catcat-native2.svg)
[Lind](https://github.com/Lind-Project/lind_project/blob/rust-graphs/graphs/catcat-lind.svg)

Some cool things from the Lind flamegraph:
1. File read is very fast! This is one of the reasons we moved to Rust
2. Lots of overhead coming from persist metadata, and we know we can optimize this further so there's room to grow.

**Comment by rennergade** [comment]
Trying the above again as a start with `cat` fixed and the other improvements made over the last ~month:

`cat 1gbfile | cat > new1gbfile`

Native (no prev file): 410 ms
Native (prev file): 2500 ms
Lind (both): 590 ms
Lind (redirect/filewrite inside of Lind): 46500 ms

I still haven't been able to pin down the native filewrite discrepancy, but Lind is now much closer to both when the redirect is outside of Lind.

When writing to a file inside of Lind it's very very slow. This is due to constant persist_metadatas. There should be ways to improve that dramatically.

**Comment by rennergade** [comment]
I also tried doing the larger grep based pipeline I was doing a year ago:
```grep '^.\{6\}$' words.txt | tr '[:upper:]' '[:lower:]' | sed 's/[abc]/2/g; s/[def]/3/g; s/[ghi]/4/g; s/[jkl]/5/g; s/[mno]/6/g; s/[pqrs]/7/g; s/[tuv]/8/g; s/[wxyz]/9/g'  | uniq | wc -l```

Times here:
Native: 680 ms
Lind: 2100 ms

I reduced this until I just had a script running grep:
Native: 670 ms ([Flamegraph](https://github.com/Lind-Project/lind_project/blob/rust-graphs/graphs/grep-nat.svg))
Lind: 1950 ms ([Flamegraph](https://github.com/Lind-Project/lind_project/blob/rust-graphs/graphs/grep-lind.svg))

So all the overhead is all coming from grep... I noticed in the FlameGraph that Lind was showing a decent portion of the graph in Wait. I then ran just the grep program through Lind (not in a script) and got:
Lind (grep, no bash): 1875 ms ([FlameGraph](https://raw.githubusercontent.com/Lind-Project/lind_project/rust-graphs/graphs/grep-lind-iso.svg)) so this actually isn't something wrong with wait(), and is all userspace.

It looks like the grep program while doing regex on a 1GB file is around 300% slower than native, which is a disappointing finding.

**Comment by rennergade** [comment]
I've been trying to figure out programs that would work well for our testing here.

As noted above, grep is slow. I used perf to find slowpoints and then looked at those addresses w/ addr2line, both were LOC that did pointer arithmetic in a tight loop. [This is somewhat noted as a slow point with NaCl.](https://theory.stanford.edu/~aiken/publications/papers/asplos17.pdf)

I looked into other standard programs. `tar` uses a smaller buffer size and streams data, but perf shows that it was only 2.5% read()/write() calls when zipping a 1GB file.

Looking for programs that are dominated by IO, `dd` is a great use case, copying a 1GB infile to an outfile showed read() as 25% of the perf time and 50%. 

Initial tests with `dd` seem promising, but had to do it hack-ily to get any results because of the metadata slowness I mentioned in the new issue.
